packages/flutter/lib/src/material/navigation_drawer.dart

```diff
@@ -0,0 +1,416 @@
+/// primary destinations in an app.
+///
+/// The style for the icons and text are not affected by parent
+/// [DefaultTextStyle]s or [IconTheme]s but rather controlled by parameters or
+/// the [NavigationDrawerThemeData].
+///
+/// This widget holds a collection of destinations ([NavigationDrawerDestination]s) and
+/// customized widgets, e.g. customized headlines, section titles, dividers.
+///
+/// Only [NavigationDrawerDestination] is counted in index.
+///
+/// {@tool dartpad}
+/// This example shows a [NavigationDrawer] as it is used within a [Scaffold]
+/// widget. The [NavigationDrawer] has headline widget, divider widget and three
+/// [NavigationDrawerDestination] widgets. The initial [selectedIndex] is set to index
+/// 0. The [onDestinationSelected] callback changes the selected item's index and displays
+/// a corresponding widget in the body of the [Scaffold].
+///
+/// ** See code in examples/api/lib/material/navigation_drawer/navigation_drawer.0.dart **
+/// {@end-tool}
+///
+/// See also:
+///
+///  * [Scaffold.drawer], where one specifies a [Drawer] so that it can be
+///    shown.
+///  * [Scaffold.of], to obtain the current [ScaffoldState], which manages the
+///    display and animation of the drawer.
+///  * [ScaffoldState.openDrawer], which displays its [Drawer], if any.
+///  * <https://material.io/design/components/navigation-drawer.html>
+class NavigationDrawer extends StatelessWidget {
+  /// Creates a Material 3 Navigation Drawer component.
+  const NavigationDrawer({
+    super.key,
+    required this.children,
+    this.backgroundColor,
+    this.surfaceTintColor,
+    this.elevation,
+    this.onDestinationSelected,
+    this.selectedIndex = 0,
+  });
+
+  /// Sets the color of the [Material] that holds all of the [NavigationDrawer]'s
+  /// contents.
+  ///
+  /// If this is null, then [NavigationDrawerThemeData.backgroundColor] is used.
+  /// If that is also null, then it falls back to [Material]'s default.
+  final Color? backgroundColor;
+
+  /// Sets the color of the [Material] that holds all of the [NavigationDrawer]'s
+  /// contents.
+  ///
+  /// If this is null, then [NavigationDrawerThemeData.surfaceTintColor] is used.
+  /// If that is also null, then it falls back to [Material]'s default.
+  final Color? surfaceTintColor;
+
+  /// The elevation of the [NavigationDrawer] itself.
+  ///
+  /// If null, [NavigationDrawerThemeData.elevation] is used. If that
+  /// is also null, it will be 1.0.
+  final double? elevation;
+
+  /// Defines the appearance of the items within the navigation drawer.
+  ///
+  /// The list contains [NavigationDrawerDestination] widgets and/or customized
+  /// widgets like headlines and dividers.
+  final List<Widget> children;
+
+  /// The index into destinations for the current selected
+  /// [NavigationDrawerDestination] or null if no destination is selected.
+  ///
+  /// Only [NavigationDrawerDestination] is counted in index.
```

Comment:
Do we have any bounds check on this to make sure it is a valid index in the constructor? We should at least comment on that here.

Reply:
If the index is out of the boundary, then no destination is selected and all destinations will be displayed in unselected style. Updated in comment too.